### ISSUE TITLE
Add a percentile function to tensorlib

### ISSUE THREAD

**Issue by matthewfeickert** [scikit-hep/pyhf#815]
# Description

First brought up in PR #790, there should be a percentile function added to the tensor backends that adheres to [NumPy's `percentile`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.percentile.html). For reference for the other backnds:

- [`tfp.stats.percentile`](https://www.tensorflow.org/probability/api_docs/python/tfp/stats/percentile)
- PyTorch doesn't have anything yet, but it is part of the NumPy functionality request rollup ([PyTorch Issue 38349](https://github.com/pytorch/pytorch/issues/38349)) and should be implemented in [PyTorch PR 39417](https://github.com/pytorch/pytorch/pull/39417).
- [`jax.numpy.percentile`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.percentile.html)

Currently PyTorch doesn't support anything that matches these other `percentile` implimentations. The [`torch.distributions.distribution.icdf`](https://pytorch.org/docs/stable/distributions.html#torch.distributions.distribution.Distribution.icdf) gives the

> the inverse cumulative density/mass function evaluated at value

which is what is desired as [`np.percentile`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.percentile.html) 

> compute[s] the q-th percentile of the data along the specified axis

but the `icdf` is a distributions method, and not a tensor method.

**Comment by matthewfeickert** [comment]
There is a [Gist of a function](https://gist.github.com/spezold/42a451682422beb42bc43ad0c0967a30) that can give something similar to `np.percentile` for PyTorch, however we want to make sure that we have the ability to have interpolation schemes like the other implementations.

**Comment by matthewfeickert** [comment]
It is also worth noting here that [TensorFlow](https://www.tensorflow.org/probability/api_docs/python/tfp/stats/percentile) has different behavior than [NumPy](https://docs.scipy.org/doc/numpy/reference/generated/numpy.percentile.html) and JAX when it comes to the "nearest" interpolation scheme.

```python
>>> import pyhf
>>> import numpy as np
>>> import tensorflow_probability as tfp
>>> pyhf.set_backend("numpy")
>>> a = pyhf.tensorlib.astensor([[10, 7, 4], [3, 2, 1]])
>>> np.percentile(a, 50, interpolation="nearest")
3.0
>>> pyhf.set_backend("tensorflow")
>>> a = pyhf.tensorlib.astensor([[10, 7, 4], [3, 2, 1]])
>>> print(tfp.stats.percentile(a, 50, interpolation="nearest"))
tf.Tensor(4.0, shape=(), dtype=float32)
```

Even though they both claim that

> This optional parameter specifies the interpolation method to use when the desired percentile lies between two data points `i < j`:
...
> ‘nearest’: `i` or `j`, whichever is nearest.

So this will need to get debugged/understood to give a consistent answer.

**Comment by matthewfeickert** [comment]
Given [my answer to the above Stack Overflow question](https://stackoverflow.com/a/61035376/8931942), I'm going to open an Issue on TensorFlow Probability's repo RE: the ordering they advertise in the docstring.

**Comment by matthewfeickert** [comment]
@lukasheinrich @kratsg Given that at the moment we're using the "linear" interpolation option (the default in NumPy) do you have any objections to me opening a PR for this Issue that could be used in PR #790 that would temporarily introduce a NumPy standin for PyTorch?

**Comment by matthewfeickert** [comment]
The next release of JAX should have all of the interpolation options in it, thanks to @jakevdp being a real MVP with [JAX PR 2610](https://github.com/google/jax/pull/2610).

**Comment by kratsg** [comment]
As tensorflow/probability#912 is merged in -- we can start figuring out if tfp has a release supporting this?

**Comment by matthewfeickert** [comment]
> we can start figuring out if tfp has a release supporting this?

It didn't make it in time for [`tfp` `v0.10.0`](https://github.com/tensorflow/probability/releases/tag/v0.10.0), but looking at [their PyPI release history](https://pypi.org/project/tensorflow-probability/0.10.0/#history) there might be a release with it in by the end of Summer.

**Comment by matthewfeickert** [comment]
In a `python:3.8-slim` Docker image

```
$ docker run --rm -it python:3.8-slim /bin/bash
:/# python -m pip install "tensorflow>=2.3.0-rc1"
:/# python -m pip install "tensorflow-probability>=0.11.0-rc0"
:/# python
Python 3.8.2 (default, Apr 23 2020, 14:32:57) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> import tensorflow_probability as tfp
>>> tf.__version__
'2.3.0-rc1'
>>> tfp.__version__
'0.11.0-rc0'
>>> 
>>> a = tf.convert_to_tensor([[10, 7, 4], [3, 2, 1]], dtype='float64')
>>> print(tfp.stats.percentile(a, 50, interpolation='linear'))
tf.Tensor(3.5, shape=(), dtype=float64)
>>> print(tfp.stats.percentile(a, 50, interpolation='nearest'))
tf.Tensor(3.0, shape=(), dtype=float64)
>>> print(tfp.stats.percentile(a, 50, interpolation='lower'))
tf.Tensor(3.0, shape=(), dtype=float64)
>>> print(tfp.stats.percentile(a, 50, interpolation='midpoint'))
tf.Tensor(3.5, shape=(), dtype=float64)
>>> print(tfp.stats.percentile(a, 50, interpolation='higher'))
tf.Tensor(4.0, shape=(), dtype=float64)

```

confirms that https://github.com/tensorflow/probability/pull/912 has made it into `tfp` `v0.11` :+1:.

**Comment by matthewfeickert** [comment]
https://github.com/pytorch/pytorch/pull/39417 has been merged in, so hopefully we see something in the [next release of PyTorch this Summer](https://pypi.org/project/torch/#history).

**Comment by matthewfeickert** [comment]
`quantile` was not added to [`torch` `v1.6.0`](https://github.com/pytorch/pytorch/releases/tag/v1.6.0) so this might take longer.

**Comment by lukasheinrich** [comment]
this is in the 1.7.0 RC2 https://github.com/pytorch/pytorch/blob/v1.7.0-rc2/docs/source/torch.rst

**Comment by matthewfeickert** [comment]
- As of today [`quantile`](https://pytorch.org/docs/1.7.0/generated/torch.quantile.html#torch.quantile) is in [`torch` `v1.7.0`](https://github.com/pytorch/pytorch/releases/tag/v1.7.0) but lacks support for interpolation methods other than `linear`.

- The fix for TFP is in `v0.11` but that requires PR #1001 and the resolution of Issue #293.

**Comment by matthewfeickert** [comment]
> * As of today [`quantile`](https://pytorch.org/docs/1.7.0/generated/torch.quantile.html#torch.quantile) is in [`torch` `v1.7.0`](https://github.com/pytorch/pytorch/releases/tag/v1.7.0) but lacks support for interpolation methods other than `linear`.

This is being worked on now and tracked in https://github.com/pytorch/pytorch/issues/48523

**Comment by matthewfeickert** [comment]
Today (2021-03-04) [`torch` `v1.8.0`](https://github.com/pytorch/pytorch/releases/tag/v1.8.0) was released, but it still lacks interpolation methods for [`quantile`](https://pytorch.org/docs/1.8.0/generated/torch.quantile.html#torch.quantile). Support was originally added in https://github.com/pytorch/pytorch/pull/48711 but was reverted as

> This PR actually breaks BC compatibility of the python API as well. Before this PR, one could call `torch.quantile(samples, quantiles, 0, False)` but now this will not run because of the missing interpolation parameter.

It is being worked on in https://github.com/pytorch/pytorch/pull/49267 but has not been completed for `v1.8.0`.

**Comment by matthewfeickert** [comment]
[`torch` `v1.8.1`](https://github.com/pytorch/pytorch/releases/tag/v1.8.1) was released on 2021-03-25, and even though https://github.com/pytorch/pytorch/pull/49267 has been merged, it is still not in a release yet. So hopefully it will be in the next one.

**Comment by matthewfeickert** [comment]
https://github.com/pytorch/pytorch/pull/59397 exists, so hopefully this means that interpolation options will be out in `torch` `v1.9.1`.

**Comment by matthewfeickert** [comment]
[`torch` `v1.9.1` was released on 2021-09-20](https://pypi.org/project/torch/1.9.1/#history) and https://github.com/pytorch/pytorch/pull/59397 has still not been merged, so maybe we'll get it in `v1.9.2`.

**Comment by matthewfeickert** [comment]
I've checked in again on https://github.com/pytorch/pytorch/pull/59397#issuecomment-965856806 but I'm not really sure if we'll see this an interpolation option in `torch` in `v1.10.1`, so we might want to implement raise not implemented errors for the time being so that PR #817 can finally move forward. :/
